util/tracecpu/collector.go

```diff
@@ -0,0 +1,333 @@
+	maxSQLNum int
+
+	// normalizedSQLMap is an map, whose keys are SQL digest strings and values are normalized SQL strings
+	normalizedSQLMap map[string]string
+
+	// normalizedPlanMap is a plan version of normalizedSQLMap
+	// this should only be set from the dedicated worker
+	normalizedPlanMap map[string]string
+
+	planRegisterChan chan *planRegisterJob
+
+	// current tidb-server instance ID
+	instanceID string
+
+	agentGRPCAddress string
+
+	quit chan struct{}
+}
+
+// TopSQLCollectorConfig is the config for TopSQLCollector
+type TopSQLCollectorConfig struct {
+	PlanBinaryDecoder   planBinaryDecodeFunc
+	MaxSQLNum           int
+	SendToAgentInterval time.Duration
+	AgentGRPCAddress    string
+	InstanceID          string
+}
+
+func encodeCacheKey(sqlDigest, planDigest string) string {
+	return sqlDigest + "-" + planDigest
+}
+
+func decodeCacheKey(key string) (string, string) {
+	split := strings.Split(key, "-")
+	sqlDigest := split[0]
+	PlanDigest := split[1]
+	return sqlDigest, PlanDigest
+}
+
+func newAgentClient(addr string, sendingTimeout time.Duration) (*grpc.ClientConn, tipb.TopSQLAgent_CollectCPUTimeClient, context.CancelFunc, error) {
+	dialCtx, cancel := context.WithTimeout(context.TODO(), time.Second)
+	defer cancel()
+	conn, err := grpc.DialContext(dialCtx, addr, grpc.WithInsecure())
+	if err != nil {
+		return nil, nil, nil, err
+	}
+	client := tipb.NewTopSQLAgentClient(conn)
+	ctx, cancel := context.WithTimeout(context.TODO(), sendingTimeout)
+	stream, err := client.CollectCPUTime(ctx)
+	if err != nil {
+		cancel()
+		return nil, nil, nil, err
+	}
+	return conn, stream, cancel, nil
+}
+
+// NewTopSQLCollector creates a new TopSQL struct
+//
+// planBinaryDecoder is a decoding function which will be called asynchronously to decode the plan binary to string
+// maxSQLNum is the maximum SQL and plan number, which will restrict the memory usage of the internal LFU cache
+func NewTopSQLCollector(config *TopSQLCollectorConfig) *TopSQLCollector {
+	normalizedSQLMap := make(map[string]string)
+	normalizedPlanMap := make(map[string]string)
+	planRegisterChan := make(chan *planRegisterJob, 10)
+	topSQLMap := make(map[string]*TopSQLDataPoints)
+
+	ts := &TopSQLCollector{
+		topSQLMap:         topSQLMap,
+		maxSQLNum:         config.MaxSQLNum,
+		normalizedSQLMap:  normalizedSQLMap,
+		normalizedPlanMap: normalizedPlanMap,
+		planRegisterChan:  planRegisterChan,
+		agentGRPCAddress:  config.AgentGRPCAddress,
+		instanceID:        config.InstanceID,
+		quit:              make(chan struct{}),
+	}
+
+	go ts.registerNormalizedPlanWorker(config.PlanBinaryDecoder)
+
+	go ts.sendToAgentWorker(config.SendToAgentInterval)
+
+	return ts
+}
+
+// Collect uses a hashmap to store records in every minute, and evict every minute.
+// This function can be run in parallel with snapshot, so we should protect the map operations with a mutex.
+func (ts *TopSQLCollector) Collect(timestamp uint64, records []TopSQLRecord) {
+	for _, record := range records {
```

Comment:
According to the protocol seems that we should keep Collect to return as fast as possible. So that maybe we need a process queue?